### ISSUE TITLE
Use of @metadata in 2.0

### ISSUE THREAD

**Issue by untergeek** [elastic/logstash#3694]
Based on https://github.com/logstash-plugins/logstash-output-elasticsearch/issues/218, and other similar requests, I propose that we start putting much of the metadata in inputs into the `@metadata` field in Logstash 2.0.  This may allow for a much cleaner output for many users.

Additionally, we could have a flag (within the input block, perhaps) that, if set, puts those fields into the event, as with 1.x.

Doing this could also pave the way to using `@metadata` more for 2.0 plugins, which is a good thing, I think.

**Comment by talevy** [comment]
I agree, I think it is sometimes dangerous to have arbitrarily named fields be "keyword" fields in only specific plugins. Keeping these special fields within the `@metadata` namespace will allow for more flexibility in setting these special fields like `type` in the ES output plugin.

**Comment by jordansissel** [comment]
I am in favor of this change. We'll need to figure out how to handle backwards compatibility/upgrading, though. Because of this schema change, folks doing, for example, a terms agg on `host` will miss all events with `@metadata.host`.
